#!/bin/bash
# This is a bash script that generates the table for number of orthologs present in each species. 
# It takes the output generated by "orthomclMclToGroups" and converts ids to numbers.
# by default it prints to stdout

# Arun Seetharam <aseetharam@purdue.edu>

scriptName="${0##*/}"

function printUsage() {
    cat <<EOF

Synopsis

    $scriptName [-h | --help] input_file

Description

    Generates the count table from the orthologous group ids file generated by "orthomclMclToGroups"
    The count table gives gene copy number in all species for each othologous group.
    
    input_file 
        Input file should contain orthologous group and IDs
        This file has to be generated by "orthomclMclToGroups" command

    -h, --help 
        Brings up this help page
 
Author

    Arun Seetharam, Bioinformatics Core, Purdue University.
    aseetharam@purdue.edu




EOF
}

if [ $# -lt 1 ] ; then
    printUsage
    exit 1
fi

while getopts ':h:' option; do
  case "$option" in
    h) printUsage
       exit
       ;;
    help) printUsage
       exit
       ;;
  esac
done

file=${1};
sed -e 's/ /" /g' -e 's/|/ "/g' -e 's/$/"/g' -e 's/:"/ /g' ${file} > ${file}.temp # separate gene ids from the species identifier
names=`head -n 500 ${file}.temp | tr -s " " "\n" |sed '/^".*/d'|sed '/^OG.*/d'| sort |uniq |tr -s "\n" " "; echo ""` #array of all species names
echo -en "OG_name\t" #print the header line
for name in ${names[@]}; do
echo -en "$name\t";
done
echo "";
while read line; do #count frequency
ogroup=$(echo $line|cut -d " " -f 1 );
echo -en "$ogroup\t";
for name2 in ${names[@]}; do
freq=`echo $line | awk -F "$name2" '{print NF-1}'`;
echo -ne "$freq\t";
done
echo "";
done<${file}.temp;
rm ${file}.temp; #delete temp file

